src/frontend/src/expr/type_inference/cast.rs

```diff
@@ -165,6 +168,75 @@ pub fn cast_map_array() -> Vec<(DataTypeName, DataTypeName, CastContext)> {
         .collect_vec()
 }
 
+#[derive(Clone, Debug)]
+pub struct CastSig {
+    pub from_type: DataTypeName,
+    pub to_type: DataTypeName,
+    pub context: CastContext,
+}
+
+/// The context a cast operation is invoked in. An implicit cast operation is allowed in a context
+/// that allows explicit casts, but not vice versa. See details in
+/// [PG](https://www.postgresql.org/docs/current/catalog-pg-cast.html).
+#[derive(Clone, Copy, Debug, Display, Eq, Ord, PartialEq, PartialOrd)]
+pub enum CastContext {
+    #[display("i")]
+    Implicit,
+    #[display("a")]
+    Assign,
+    #[display("e")]
+    Explicit,
+}
+
+pub type CastMap = BTreeMap<(DataTypeName, DataTypeName), CastContext>;
+
+pub fn cast_sigs() -> impl Iterator<Item = CastSig> {
+    CAST_MAP
+        .iter()
+        .map(|((from_type, to_type), context)| CastSig {
+            from_type: *from_type,
+            to_type: *to_type,
+            context: *context,
+        })
+}
+
+pub static CAST_MAP: LazyLock<CastMap> = LazyLock::new(|| {
+    use DataTypeName::*;
+    const CAST_TABLE: &[(&str, DataTypeName)] = &[
+        // 123456789ABCDEF
+        ("  e            a", Boolean),     // 0
+        ("  iiiiii       a", Int16),       // 1
+        ("ea iiiii       a", Int32),       // 2
+        (" aa iiii       a", Int64),       // 3
+        (" aaa ii        a", Decimal),     // 4
+        (" aaaa i        a", Float32),     // 5
+        (" aaaaa         a", Float64),     // 6
+        ("      e        a", Int256),      // 7
+        ("          ii   a", Date),        // 8
+        ("            i  a", Time),        // 9
+        ("        aa i   a", Timestamp),   // A
+        ("        aaa    a", Timestamptz), // B
+        ("         a     a", Interval),    // C
+        ("eeeeeee        a", Jsonb),       // D
+        ("               a", Bytea),       // E
+        ("eeeeeeeeeeeeeee ", Varchar),     // F
+    ];
```

Comment:
Personally I feel this table is comparably less intuitive than the previous groups:
* 3 implicit/assign sequences: int16..float64, date..timestamptz, time..interval
* any->varchar is assign and varchar->any is explicit
* jsonb->bool/number is explicit
* misc: explicit int32<->bool, assign timestamp/timestamptz->time
* int256

Is it purely different style preference or do you have other motivations for the change?

Reply:
Just a style preference. I think a table can define the case rules in a clear and compact way, like our [documentation](https://docs.risingwave.com/docs/current/data-type-casting/). What do you think if we explain these rules in the comments?

Reply:
Sure let's at least add them as comments. It is not affecting other parts to switch between styles as well.